lib/engine-addon.js

```diff
@@ -330,36 +333,27 @@ module.exports = {
       let invokeArguments = args || [];
       let hostAddons = this.ancestorHostAddons();
 
-      return this.addons
-        .map(addon => {
+      if (methodName === 'treeFor') {
+        deeplyNonDuplicatedAddon(hostAddons, this, invokeArguments[0]);
```

Comment:
Although in most cases this should be fine, I would like to recommend we put this new behavior behind a flag (off by default), and release as a minor version bump. After some time, we can swap the default to be true, and do a major version bump. Then at some future point in time, if all seems well, we can drop the flag entirely.

Reply:
If the flag is off, should it fall back to the old non deeply deduped implementation?

Reply:
Yes